File: AIM/AIM/Data/Gateways/XMLOnAirInfoGateway.m
//
//  XMLOnAirInfoGateway.m
//  AIM
//
//  Created by Oskar Szydlowski on 06.03.2018.
//  Copyright © 2018 Oskar Szydlowski. All rights reserved.
//

#import "XMLOnAirInfoGateway.h"

@implementation XMLOnAirInfoGateway

- (id)init {
    self = [super init];
    if (self != nil) {
        self.parser = [[NSXMLParser new] initWithContentsOfURL: [[NSURL alloc] initWithString: @"http://aim.appdata.abc.net.au.edgesuite.net/data/abc/triplej/onair.xml"]];
        self.parser.delegate = self;
        self.tracks = [[NSMutableArray alloc] init];
        self.stationDisplayTime = [NSString new];
    };
    return self;
}

- (void)fetchOnAirInfo:(OnAirInfoCompletion)completion {
    self.onAirInfoCompletion = completion;
    [self.parser parse];
}


// MARK: - XMLParserDelegate

- (void)parser:(NSXMLParser *)parser didStartElement:(NSString *)elementName namespaceURI:(NSString *)namespaceURI qualifiedName:(NSString *)qName attributes:(NSDictionary<NSString *,NSString *> *)attributeDict {
    if ([elementName  isEqual: @"epgItem"]) {
        self.stationInfo = [[StationInfo new] initWithAttributesDictionary:attributeDict];
    } else if ([elementName  isEqual: @"customField"]) {
        if (attributeDict[@"name"] != nil) {
            if ([attributeDict[@"name"]  isEqual: @"displayTime"]) {
                self.stationDisplayTime = attributeDict[@"value"];
            } else if ([attributeDict[@"name"]  isEqual: @"image320"]) {
                self.stationImageURL = [[NSURL alloc] initWithString:attributeDict[@"value"]];
            }
        }
    } else if ([elementName  isEqual: @"playoutItem"]) {
        self.track = [[Track new] initWithAttributesDictionary:attributeDict];
    }
}

- (void)parser:(NSXMLParser *)parser didEndElement:(NSString *)elementName namespaceURI:(NSString *)namespaceURI qualifiedName:(NSString *)qName; {
    if ([elementName  isEqual: @"playoutItem"]) {
        if (self.track != nil) {
            [self.tracks addObject:self.track];
        }
    }
}

- (void)parserDidEndDocument:(NSXMLParser *)parser; {
    if (self.stationInfo == nil || self.stationDisplayTime == nil || self.stationImageURL == nil) {
        self.onAirInfo = [[OnAirInfo new] initWithStationInfo:self.stationInfo tracks:self.tracks];
        self.onAirInfoCompletion(self.onAirInfo);
        return;

    }
    self.stationInfo = [[StationInfo new] initWithName:self.stationInfo.name description:self.stationInfo.description time:self.stationInfo.time duration: self.stationInfo.duration presenter:self.stationInfo.presenter displayTime:self.stationDisplayTime imageURL:self.stationImageURL];
    self.onAirInfo = [[OnAirInfo new] initWithStationInfo:self.stationInfo tracks: self.tracks];
    self.onAirInfoCompletion(self.onAirInfo);
}

@end

File: AIM/AIM/Data/Gateways/XMLOnAirInfoGateway.swift
//
//  XMLOnAirInfoGateway.swift
//  AIM
//
//  Created by Oskar Szydlowski on 01.03.2018.
//  Copyright © 2018 Oskar Szydlowski. All rights reserved.
//

import Foundation

typealias OnAirInfoCompletion = (OnAirInfo) -> ()

class XMLOnAirInfoGateway: NSObject, XMLParserDelegate {
    
    private let onAirInfoFeedURL = URL(string: "http://aim.appdata.abc.net.au.edgesuite.net/data/abc/triplej/onair.xml")!
    private var parser: XMLParser
    
    private enum XMLKeys: String {
        case stationInfo = "epgItem"
        case customField = "customField"
        case displayTime = "displayTime"
        case image = "image320"
        case name = "name"
        case value = "value"
        case track = "playoutItem"
    }
    
    private var stationDisplayTime: String?
    private var stationImageURL: URL?
    private var stationInfo: StationInfo?
    private var tracks = [Track]()
    private var track: Track?
    
    private var parserCompletion: OnAirInfoCompletion?
    
    override init() {
        self.parser = XMLParser(contentsOf: onAirInfoFeedURL)!
        super.init()
        parser.delegate = self
    }
    
    func parse(completion: @escaping OnAirInfoCompletion) {
        parserCompletion = completion
        parser.parse()
    }
    
    // MARK: - XMLParserDelegate
    
    func parser(_ parser: XMLParser, didStartElement elementName: String, namespaceURI: String?, qualifiedName qName: String?, attributes attributeDict: [String : String] = [:]) {
        if elementName == XMLKeys.stationInfo.rawValue {
            stationInfo = StationInfo.fromAttributeDict(attributeDict: attributeDict)
        } else if elementName == XMLKeys.customField.rawValue {
            if let attributeName = attributeDict[XMLKeys.name.rawValue] {
                if attributeName == XMLKeys.displayTime.rawValue {
                    stationDisplayTime = attributeDict[XMLKeys.value.rawValue]
                } else if attributeName == XMLKeys.image.rawValue {
                    if let imageURLString = attributeDict[XMLKeys.value.rawValue], let imageURL = URL(string: imageURLString) {
                        stationImageURL = imageURL
                    }
                }
            }
        } else if elementName == XMLKeys.track.rawValue {
            track = Track.fromAttributeDict(attributeDict: attributeDict)
        }
    }
    
    func parser(_ parser: XMLParser, didEndElement elementName: String, namespaceURI: String?, qualifiedName qName: String?) {
        if elementName == XMLKeys.track.rawValue, let track = track {
            tracks.append(track)
        }
    }

    func parserDidEndDocument(_ parser: XMLParser) {
        guard
            let stationInfo = stationInfo,
            let stationDisplayTime = stationDisplayTime,
            let stationImageURL = stationImageURL
        else {
            let onAirInfo = OnAirInfo(stationInfo: nil, tracks: tracks)
            parserCompletion?(onAirInfo)
            return
        }
                
        let onAirInfo = OnAirInfo(
            stationInfo:
                StationInfo(
                    name: stationInfo.name,
                    description: stationInfo.description,
                    time: stationInfo.time,
                    duration: stationInfo.duration,
                    presenter: stationInfo.presenter,
                    displayTime: stationDisplayTime,
                    imageURL: stationImageURL
                ),
            tracks: tracks
        )
        parserCompletion?(onAirInfo)
    }
}
